### PR TITLE
Updated FormatMoneyUseCase

### LINKED ISSUE #3642: [FEATURE] Add KDoc And Crypto Formatting
### Please confirm the following:

- [X] I've checked [the current issues](https://github.com/Ivy-Apps/ivy-wallet/issues) for duplicate issues.
- [X] I've requested a single (only one) feature/change in this issue. It complies with the One Request Per GitHub Issue (ORPGI) rule.
- [X] My issue is well-defined and describes how it should be implemented from UI/UX perspective.

### What do you want to be added or improved?

 In [FormatMoneyUseCase](https://github.com/Ivy-Apps/ivy-wallet/blob/main/shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt) Do :

1. Add support for crypto formatting
2. KDoc for better readability

### Why do you need it?

-  Additional feature + better understanding

### How do you imagine it?

_No response_

### PR DESCRIPTION
**Pull request by shamim-emon**
## Pull request (PR) checklist

Please check if your pull request fulfills the following requirements:

- [x] I've read the [Contribution Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/CONTRIBUTING.md) and my PR doesn't break the rules.
- [x] I've read and understand the [Developer Guidelines](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/Guidelines.md).
- [x] I confirm that I've run the code locally and everything works as expected.
- [x] My PR includes only the necessary changes to fix the issue (i.e., no unnecessary files or lines of code are changed).
- [ ] 🎬 I've attached a **screen recording** of using the new code to the next paragraph (if applicable).

## What's changed?

Describe with a few bullets **what's new:**
I've added below in FormatMoneyUseCase
- Crypto formatting
- KDocumentation

## Risk factors

**What may go wrong if we merge your PR?**
- N/A

**In what cases won't your code work?**
- N/A

## Does this PR close any GitHub issues? (do not delete)

- Closes #3642 

## Troubleshooting GitHub Actions (CI) failures ❌
Pull request checks failing? Read our [CI Troubleshooting guide](https://github.com/Ivy-Apps/ivy-wallet/blob/main/docs/CI-Troubleshooting.md).

### CODE REVIEW

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt [hunk 1]
```diff
@@ -23,25 +32,69 @@ class FormatMoneyUseCase @Inject constructor(
...
+    /**
+     * Formats a currency or cryptocurrency amount based on the input parameters.
+     *
+     * @param value The numeric value to format.
+     * @param shortenAmount Flag to indicate if the amount should be shortened (e.g., "1k" for 1,000).
+     * @param isCrypto Flag to indicate if the value is a cryptocurrency, enabling up to 9 decimal places.
+     * @return The formatted string representation of the value.
+     */
+    suspend fun format(value: Double, shortenAmount: Boolean, isCrypto: Boolean = false): String {
+        val result = if (isCrypto) {
```
Comment (ILIYANGERMANOV) [inline shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt:47]: You can directly `return`, no need for result var

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt [hunk 2]
```diff
@@ -23,25 +32,69 @@ class FormatMoneyUseCase @Inject constructor(
...
+    /**
+     * Formats a cryptocurrency value with up to 9 decimal places, removing unnecessary trailing zeros.
+     *
+     * @param value The cryptocurrency value to format.
+     * @return The formatted cryptocurrency value as a string.
+     */
+    private fun formatCrypto(value: Double): String {
+        val result = cryptoFormatter.format(value)
+        return when {
+            result.lastOrNull() == localDecimalSeparator().firstOrNull() -> {
```
Comment (ILIYANGERMANOV) [inline shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt:79]: I don't like this logic - can we just fix the formatter?

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt [hunk 3]
```diff
@@ -23,25 +32,69 @@ class FormatMoneyUseCase @Inject constructor(
     private val withoutDecimalFormatter = DecimalFormat("###,###", DecimalFormatSymbols(locale))
     private val withDecimalFormatter = DecimalFormat("###,###.00", DecimalFormatSymbols(locale))
     private val shortenAmountFormatter = DecimalFormat("###,###.##", DecimalFormatSymbols(locale))
+    private val cryptoFormatter =
+        DecimalFormat("###,###,##0.${"0".repeat(9)}", DecimalFormatSymbols(locale))
```
Comment (ILIYANGERMANOV) [inline shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt:36]: Can you try fixing the formatter by `"#".repeat(9)`

File: shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt [hunk 4]
```diff
@@ -24,79 +24,122 @@ class FormatMoneyUseCaseTest {
...
+            locale = Locale.ENGLISH,
+            expectedOutput = "123,456.000000000"
+        ),
+        ENG_HIDE_DECIMAL_CRYPTO(
+            amount = 123_456.0,
+            showDecimal = false,
+            shortenAmount = false,
+            isCrypto = true,
+            locale = Locale.ENGLISH,
+            expectedOutput = "123,456.000000000"
```
Comment (ILIYANGERMANOV) [inline shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt:125]: We don't expect such zeros

**Review by ILIYANGERMANOV — CHANGES_REQUESTED** [review]
Left some comments but "1,234.000000000000" isn't expected

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt (hunk 1 above)
Comment (shamim-emon) [reply shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt]: Done

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt (hunk 2 above)
Comment (shamim-emon) [reply shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt]: Removed the function, just relying on the suggested formatter now

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt (hunk 3 above)
Comment (shamim-emon) [reply shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt]: Done

File: shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt (hunk 4 above)
Comment (shamim-emon) [reply shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt]: Updated the result as-per the suggested formatter expected result

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt [hunk 5]
```diff
@@ -23,25 +32,35 @@ class FormatMoneyUseCase @Inject constructor(
     private val withoutDecimalFormatter = DecimalFormat("###,###", DecimalFormatSymbols(locale))
     private val withDecimalFormatter = DecimalFormat("###,###.00", DecimalFormatSymbols(locale))
     private val shortenAmountFormatter = DecimalFormat("###,###.##", DecimalFormatSymbols(locale))
+    private val cryptoFormatter = DecimalFormat("#".repeat(9), DecimalFormatSymbols(locale))
```
Comment (ILIYANGERMANOV) [inline shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt:35]: Ah, I mean we need to support showing any arbitrary crypto with up to 9 decimals precision without showing unnecessary trailing digits. The correct pattern should be:
- `"###,###." + "#".repeat(9)`

File: shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt [hunk 6]
```diff
@@ -24,79 +24,122 @@ class FormatMoneyUseCaseTest {
...
+        ),
+        ENG_HIDE_DECIMAL_CRYPTO(
+            amount = 123_456.0,
+            showDecimal = false,
+            shortenAmount = false,
+            isCrypto = true,
+            locale = Locale.ENGLISH,
+            expectedOutput = "123456"
+        ),
+        GERMAN_SHOW_DECIMAL_CRYPTO(
```
Comment (ILIYANGERMANOV) [inline shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt:127]: We need a tear case for "0.000345 BTC" for example

File: screen/edit-transaction/src/main/java/com/ivy/transaction/EditTransactionScreen.kt [hunk 7]
```diff
@@ -56,7 +56,7 @@ import com.ivy.navigation.navigation
 import com.ivy.navigation.screenScopedViewModel
 import com.ivy.ui.R
 import com.ivy.wallet.domain.data.CustomExchangeRateState
-import com.ivy.wallet.domain.data.IvyCurrency
+import com.ivy.legacy.domain.data.IvyCurrency
```
Comment (ILIYANGERMANOV) [inline screen/edit-transaction/src/main/java/com/ivy/transaction/EditTransactionScreen.kt:59]: What caused this package change?

File: shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt (hunk 5 above)
Comment (shamim-emon) [reply shared/ui/core/src/main/java/com/ivy/ui/FormatMoneyUseCase.kt]: Done

File: shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt (hunk 6 above)
Comment (shamim-emon) [reply shared/ui/core/src/test/java/com/ivy/ui/FormatMoneyUseCaseTest.kt:143]: Done

File: screen/edit-transaction/src/main/java/com/ivy/transaction/EditTransactionScreen.kt (hunk 7 above)
Comment (shamim-emon) [reply screen/edit-transaction/src/main/java/com/ivy/transaction/EditTransactionScreen.kt:59]: Package directive of `IvyCurrency` data class  was wrong.  It resides in `com.ivy.legacy.domain.data.IvyCurrency` but directive was `com.ivy.wallet.domain.data.IvyCurrency` in the data class file. I fixed it thus these changes appeared as-per correct package directive in places where it's used.

**Comment by ILIYANGERMANOV** [comment]
Thank you for your contribution to the Ivy Wallet project. We appreciate the time and effort you've invested in this pull request.

As of Nov 5th, 2024, Ivy Wallet is no longer maintained. We are closing all open pull requests as part of the project shutdown.

While we won't be merging this pull request, you're welcome to fork the repository and continue development independently under the terms of the [GPL-3.0 License](https://github.com/Ivy-Apps/ivy-wallet/blob/main/LICENSE).

For more information, please refer to the shutdown notice in our [README](https://github.com/Ivy-Apps/ivy-wallet/blob/main/README.md) file.

We apologize for any inconvenience and thank you again for your interest in improving Ivy Wallet.

**Pull request closed without merging**